src/server-core/src/lib.rs

```diff
@@ -156,52 +157,104 @@ pub async fn listen(
     Ok((handle, Box::pin(stream)))
 }
 
-/// Serves incoming TCP connections from `conns` using `server`.
-pub async fn serve<C, S>(mut conns: C, server: S)
+/// Serves incoming TCP connections from `conns` using `server`. `wait_timeout` is the time to wait
+/// after `conns` for outstanding connections to complete. Returns handles to the outstanding
+/// connections after `wait_timeout` has expired or all connections have completed.
+pub async fn serve<C, S>(mut conns: C, server: S, wait_timeout: Option<Duration>) -> JoinSet<()>
```

Comment:
Food for thought, if we start to grow more optional config params here, then a builder pattern could be nice!